src/vsr/replica.zig

```diff
@@ -8005,7 +8036,11 @@ pub fn ReplicaType(
                 self.transition_to_recovering_head();
             }
 
-            self.state_machine.open(state_machine_open_callback);
+            self.superblock.free_set_encoded.open(
+                &self.grid,
+                self.superblock.working.free_set_reference(),
+                free_set_open_callback,
+            );
```

Comment:
I'd say the free set (encoded) should go from the superblock entirely maybe? And move into grid, so that we have `grid.open` / `grid.checkpoint` 

:thinking: in https://github.com/tigerbeetle/tigerbeetle/pull/1309#discussion_r1402525289, I wrote

> maybe I find some better way to assert that (that the grid checkpoing uses correct state of the free set)...

This might actually be it? If the grid owns the free set, than checkpointing the free set and waiting for repairs is handled by the single `grid.checkpoing` function, and that can _locally_ sequence the two correctly (as opposed to the current situtation, where the replica has to know to checkpoing free set before the grid).

Reply:
Yeah, that works. At the same time maybe instead of the replica getting the `FreeSetReference` via a method on `FreeSetEncoded` (or the `Grid`), the `FreeSetReference` could be passed into the `checkpoint` callback (which is then passed into `superblock.checkpoint()`).

Reply:
Yeah, thought the same as well, but didn't do for two reasons:

* to use the same pattern as in manifest log
* to avoid requiring the replica to keep this bit of state around between callbacks: the replica goes through commit dispatch, so it can't immediately just pass the checkpoint along, it would have to store it as a field, which feels awkward. 

Perhaps in the both cases we rather need a state enum?

```
cosnt state = union(enum) {
   closed,
   at_checkpoint: CheckpointReference,
   working,
}
```

then, checkpoint reference would not be just a getter, but would actually change the state from `at_checkpoint` to `working`.

Reply:
Ah yeah.. it kind of seems like `CommitStage` should be a `union(enum)`, and the checkpoint stages would accumulate the references. But maybe that is more verbose than just doing what we do for the manifest log references :shrug: